### PR TITLE
added support for loading custom driver

### PR DESCRIPTION
**Pull request by cocoa-xu**

### CODE REVIEW

**Comment by josevalim** [comment]
Sorry @cocoa-xu, I just pushed this feature. :) However, I didn't include any tests. Maybe we can amend this?

**Comment by cocoa-xu** [comment]
Sure! I'll add some tests for it

File: test/adbc_database_test.exs [hunk 1]
```diff
@@ -33,4 +33,53 @@ defmodule Adbc.DatabaseTest do
...
+
+  describe "load custom driver" do
+    @describetag :duckdb
+
+    setup do
+      {:ok, zip_data} =
+        Adbc.Helper.download(
+          "https://github.com/duckdb/duckdb/releases/download/v0.8.1/libduckdb-linux-amd64.zip",
+          false
+        )
```
Comment (josevalim) [inline test/adbc_database_test.exs:45]: Ohhhhhh, maybe we should allow a custom path to be given to `Adbc.download_driver`? Or are the driver structures different? If it works, then we make `Adbc.download_driver` return the path to the driver, which we can then pass to `Adbc.Database`. WDYT? Or is the structure different?

File: test/adbc_database_test.exs [hunk 2]
```diff
@@ -33,4 +33,53 @@ defmodule Adbc.DatabaseTest do
       assert Exception.message(error) == "[SQLite] Unknown database option who_knows=123"
     end
   end
+
+  describe "load custom driver" do
+    @describetag :duckdb
```
Comment (josevalim) [inline test/adbc_database_test.exs:38]: Should we exclude this by default in test_helper.exs?

File: test/adbc_database_test.exs [hunk 3]
```diff
@@ -33,4 +33,53 @@ defmodule Adbc.DatabaseTest do
...
+          false
+        )
+
+      cache_path = Path.join(File.cwd!(), "libduckdb-linux-amd64.zip")
+      File.write!(cache_path, zip_data)
+
+      adbc_so_priv_dir =
+        case :os.type() do
+          {:win32, _} -> Application.app_dir(:adbc, "priv/bin")
+          _ -> Application.app_dir(:adbc, "priv/lib")
```
Comment (josevalim) [inline test/adbc_database_test.exs:53]: Let's not write to `priv`. Later we can make a release and accidentally ship the stuff in priv. A "tmp" directory would be better. :)

**Comment by cocoa-xu** [comment]
And I found that if the entry point is invalid, it seems that the program would crash in CI. But it works fine if I test in in IEx

```
iex> Adbc.Database.start_link(driver: Path.join(File.cwd!(), "libduckdb.so"), entrypoint: "entrypoint")
{:error,
 #Adbc.Error<
   message: "dlsym(entrypoint) failed: /home/cocoa/adbc/libduckdb.so: undefined symbol: entrypoint",
   vendor_code: 0,
   ...
 >}
```

File: test/adbc_database_test.exs (hunk 1 above)
Comment (cocoa-xu) [reply test/adbc_database_test.exs:45]: Yeah, it's different. For the duckdb one, it only consists 3 files, and it's using the `.dylib` extension for macOS

```
$ unzip libduckdb-linux-amd64.zip
Archive:  libduckdb-linux-amd64.zip
  inflating: libduckdb.so
  inflating: duckdb.hpp
  inflating: duckdb.h
```

However, for the official ones, they're organised as python wheels, so the actual driver is inside a directory 

```
$ unzip adbc_driver_sqlite-0.5.1-py3-none-macosx_11_0_arm64.whl
$ tree adbc_driver_sqlite-0.5.1-py3-none-macosx_11_0_arm64
.
├── adbc_driver_sqlite
│   ├── __init__.py
│   ├── _static_version.py
│   ├── _version.py
│   ├── dbapi.py
│   ├── libadbc_driver_sqlite.so
│   └── py.typed
└── adbc_driver_sqlite-0.5.1.dist-info
    ├── LICENSE.txt
    ├── METADATA
    ├── NOTICE.txt
    ├── RECORD
    ├── WHEEL
    └── top_level.txt
```

Comment (cocoa-xu) [reply test/adbc_database_test.exs:45]: Or maybe we can allow users to pass a callback function for it? This callback function will be used to indicate which files to extract from the zip file.

But in the case where the URL points to other archive formats, like `.tar.gz`, we have to let the user to proper library to extract the driver.

File: test/adbc_database_test.exs [hunk 4]
```diff
@@ -33,4 +33,49 @@ defmodule Adbc.DatabaseTest do
       assert Exception.message(error) == "[SQLite] Unknown database option who_knows=123"
     end
   end
+
+  describe "load custom driver" do
+    @describetag :duckdb
+
+    setup do
```
Comment (josevalim) [inline test/adbc_database_test.exs:40]: Should we change this to setup_all?

File: test/adbc_database_test.exs (hunk 1 above)
Comment (josevalim) [reply test/adbc_database_test.exs:45]: I see. So let's not do anything like this for now.

File: test/adbc_database_test.exs (hunk 4 above)
Comment (cocoa-xu) [reply test/adbc_database_test.exs]: ok, will do that!

File: test/adbc_database_test.exs [hunk 5]
```diff
@@ -3,6 +3,37 @@ defmodule Adbc.DatabaseTest do
...
+    {:ok, zip_data} =
+      Adbc.Helper.download(
+        "https://github.com/duckdb/duckdb/releases/download/v0.8.1/libduckdb-linux-amd64.zip",
+        false
+      )
+
+    cache_path = Path.join(File.cwd!(), "libduckdb-linux-amd64.zip")
+    File.write!(cache_path, zip_data)
+
+    tmp_dir = System.tmp_dir!()
```
Comment (josevalim) [inline test/adbc_database_test.exs:16]:
```suggestion
    tmp_dir = System.tmp_dir!()
    cache_path = Path.join(tmp_dir, "libduckdb-linux-amd64.zip")
    File.write!(cache_path, zip_data)
```

**Review by josevalim — APPROVED** [review]

**Pull request merged**